

# <span style="color:darkblue">Módulo 1 - Introducción a la programación con Python </span>

## <span style="color:darkblue">Sesión:	Sesión 2 - Características generales de Python Sintaxis</span>

### <span style="color:darkblue">Duración estimada:	3 horas</span>


## <span style="color:darkblue"> Tabla de contenidos:
 
    1- Bienvenida
    
    2- Asistencia
    
    3- Dudas de la Clase anterior
    
    4- Control de Lectura
    
    5- Objetivo
    
    6- Introducción
    
    7- Funciones Integradas
    
    8- Variables 
    
    9- Control de lectura
    
    10-Función Print()
    
    12-Control de lectura
    
    13- Comentarios
    
    14-Asistencia
    
    15 - Tarea
</span>

## Objetivo:
Comprender las reglas y convenciones que se utilizan para escribir programas del lenguaje de programación de Python, aprender estas características generales de Python te ayudará a escribir código Python de manera más efectiva y a comprender mejor los programas Python escritos por otros programadores.

### Recursos

•	Sesión 2 de Curso de Python

•	Acceso a internet

•	Python Instalado en computadora

•	Visual Studio Code

### Introducción:
Es importante comprender algunas de las características generales de la sintaxis de Python para poder dominarlo en primer lugar, Python utiliza la indentación para definir bloques de código, lo que significa que la forma en que se estructura el código es crucial para su correcta ejecución. 

Además, Python tiene una gran cantidad de tipos de datos incorporados, como números, cadenas de texto y listas, lo que lo hace versátil y adecuado para una amplia variedad de aplicaciones.

Otras características importantes incluyen el uso de estructuras de control de flujo como if/else y bucles for/while, el uso de funciones para organizar y reutilizar el código, y el uso de módulos para importar funcionalidad de otros archivos de Python. 

Por último, Python tiene una sólida gestión de excepciones, lo que permite manejar errores y problemas de manera efectiva en tiempo de ejecución.

### La línea de comandos de Python

Para probar una pequeña cantidad de código en python, a veces es más rápido y fácil no escribir el código en un archivo. Esto es posible porque Python se puede ejecutar como una línea de comando.

Escriba lo siguiente en la línea de comandos:

<img src="./images/01.png" alt="">

Que escribirá "¡Hola, mundo!" en la línea de comando:

<img src="./images/02.png" alt="">

Esa es una forma fácil de trabajar código rápido de Python desde la consola. Recordemos que en nuestro curso trabajaremos con Visual Studio Code y debido a ello tendremos en cuenta el tener una carpeta donde almacenaremos nuestro código. 


In [2]:
#funcion integrada para calcular el maximo

result = max(4, 5, 6)
result

6

In [6]:
result = 137/6
round(result, 0)

23.0

In [8]:
result = min(4,5,6,7)
result

4

### <span style="color:darkred">Python como calculadora</span>

In [9]:
3 + 5

8

In [10]:
4*5

20

Potencia de un número $3^2$

In [11]:
3**2

9

Division y nos muestra el cociente

In [20]:
8 // 5

1

División para que nos muestre el residuo

In [15]:
8 % 5

3

Calculando muchas operaciones simultaneas

In [19]:
1 + 2 -5 * 30 / 5

-27.0

In [21]:
def sumar(a,b):
    return a+b

In [22]:
sumar(3,5)

8

### <span style="color:darkred">Built-in Functions - Funciones Integradas</span>

En Python tenemos muchas funciones integradas. Las funciones integradas están disponibles globalmente para que las uses sin importar módulos ni ninguna configuración adicional. Algunas de las funciones integradas más comunes son:

```print(), len(), type(), int(), float(), str(), input(), list(), dict(), min(), max(), sum(), sorted(), open(), file(), help(), and dir().```

<img src="./images/04.png" alt="">

In [25]:
print('Hello, world!')
print("Hello, world!")

len("Hello, world!") # Empieza a contar asi index [0, 1, 2, 3, ... , 13]

Hello, world!
Hello, world!


13

In [26]:
type("Hello, world!")

str

In [27]:
10

10

In [28]:
str(10)

'10'

In [29]:
float(10)

10.0

In [30]:
input('Escribe aca la cadena de caracteres a capturar')

Escribe aca la cadena de caracteres a capturaresta es una prueba de captura de texto


'esta es una prueba de captura de texto'

In [32]:
numero = input('Escribe un numero: ')

print(type(numero))
print(numero)

Escribe un numero: 34
<class 'str'>
34


In [33]:
numero = int(input('Escribe un numero: '))

print(type(numero))
print(numero)

Escribe un numero: 34
<class 'int'>
34


In [34]:
numero = float(input('Escribe un numero: '))

print(type(numero))
print(numero)

Escribe un numero: 34
<class 'float'>
34.0


In [35]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



In [36]:
help(int)

Help on class int in module builtins:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      True if 

### Práctica: (3 minutos)

1- Definir una variable `nombre` que guarde el nombre de ustedes, `edad` que almacene la edad.

2- Imprimir `print('Mi nombre es', nombre, 'y mi edad es :',edad)`

3- encuentre la longitud de la variable `nombre`.


In [37]:
nombre = 'José Fuentes'
edad   = 38

print('Mi Nombre es: ',nombre,'y mi edad es: ',edad)

Mi Nombre es:  José Fuentes y mi edad es:  38


In [39]:
string1 = 'Hola buenos días'
stgMed  = ','
string2 = 'es un gusto saludarlos'

concatenada = string1 + stgMed + string2

print(concatenada)

Hola buenos días,es un gusto saludarlos


Las características principales que Python tiene son las siguientes: 

<img src="./images/03.png" alt="">

In [40]:
if True:
    print('Esta es una prueba de identación',3 + 4)

Esta es una prueba de identación 7


In [41]:
if True:
print('Esta es una prueba de identación',3 + 4)

IndentationError: expected an indented block (113113059.py, line 2)

In [43]:
value = 3

if value > 4:
    print('El numero es mayor que 4')
    
    if value % 2 == 0:
        print('el numero es par')
    
    else:
        print('el numero es impar')
else:
    print('el numero es menor que 4')

el numero es menor que 4


###   <span style="color:darkred">Comentarios</span>

Se pueden añadir comentarios en el código de Python utilizando el símbolo "#" al comienzo de una línea. Los comentarios son útiles para explicar el propósito de un código y hacerlo más fácil de entender.

Los objetivos principales de los comentarios son los siguientes:

   * Los comentarios se pueden usar para explicar el código de Python.
   * Los comentarios se pueden utilizar para hacer que el código sea más legible.
   * Los comentarios se pueden usar para evitar la ejecución al probar el código

Los comentarios comienzan con un # y Python representará el resto de la línea como un comentario:

### <span style="color:darkblue">Comentarios por línea</span>

In [46]:
# This is a comment

print('hola Mundo')

hola Mundo


In [48]:
print('hola Mundo')
#this is a comment

hola Mundo


En este caso el texto “This is a comment” se traducirá como comentario y no se tomará en cuenta para la ejecución de código de Python. 

Los comentarios se pueden colocar al final de una línea y Python ignorará el resto de la línea:


In [47]:
print('hola Mundo') # this is a comment

hola Mundo


Un comentario no tiene que ser texto que explique el código, también se puede usar para evitar que Python ejecute código:

In [50]:
#Esta funccion saluda a alguin dado su nombre

def saludar(nombre):
    #Esta funccion saluda a alguin dado su nombre
    mensaje = 'Hola, ' + nombre + '!'
    print(mensaje)
    
saludar('Juan')
    

Hola, Juan!


### <span style="color:darkblue">Comentarios multilínea</span>

Python realmente no tiene una sintaxis para comentarios de varias líneas.

Para agregar un comentario de varias líneas, puede insertar un #para cada línea:

In [51]:
#  linea 1: Este es un comentario multilinea
#  linea 2:Los comentarios inline proporcionan descripciones
#  linea 3: breves de variables y operaciones sencillas, y se escriben 
#  linea 4: en la misma línea que la sentencia del códig
print('Hola mundo')

Hola mundo


O, no del todo como se pretendía, puede usar una cadena de varias líneas.

Dado que Python ignorará los literales de cadena que no están asignados a una variable, puede agregar una cadena de varias líneas (comillas triples) en su código y colocar su comentario dentro de ella:


In [53]:
""" pueden escribir lo que quieran aca"""
print('hola mundo')

hola mundo


In [54]:
"""Dado que Python ignorará los literales de cadena 
que no están asignados a una variable, 
puede agregar una cadena de varias líneas 
(comillas triples) en su código y colocar su 
comentario dentro de ella:"""
print('hola Mundo')

hola Mundo


In [56]:
def saludar(nombre):
    """Esta es una funcio que da un saludo a alguien 
    si se le intruce el nombre del quieren saludar
    Quedo claro."""
    mensaje = 'Hola, ' + nombre + '!'
    print(mensaje)
    
saludar('Juan')

Hola, Juan!


### <span style="color:darkred">Variables</span>

Python no requiere declaraciones de variables. 

Una variable se crea simplemente asignando un valor a un nombre. 

Las variables pueden contener cualquier tipo de datos, incluyendo **números, cadenas de texto y objetos complejos.**

<span style="color:darkgreen">Una variable se crea en el momento en que le asigna un valor por primera vez. </span>

In [57]:
#definamos una variable
x = 5

print(x)

5


In [58]:
x = 34.45456

print(x)

34.45456


In [60]:
x = 'Hola, mundo'

print(x)

Hola, mundo


In [62]:
y = 'John Doe'
x = 4

In [63]:
x * y 

'John DoeJohn DoeJohn DoeJohn Doe'

In [64]:
5*y

'John DoeJohn DoeJohn DoeJohn DoeJohn Doe'

In [71]:
print('************************')
print('****BIENVENIDOS*********')
print('************************')

************************
****BIENVENIDOS*********
************************


In [68]:
print(21*'*')
print(5*'*'+'BIENVENIDOS'+5*'*')
print(21*'*')

*********************
*****BIENVENIDOS*****
*********************


In [74]:
x = 'john'
x = 3
3*x

9

### <span style="color:darkblue">Nombres de variables</span>

Una variable puede tener un nombre corto (como x e y) o un nombre más descriptivo (edad, nombre del coche, volumen_total). Reglas para las variables de Python:

* Un nombre de variable debe comenzar con una letra o el carácter de subrayado

* Un nombre de variable no puede comenzar con un número

* Un nombre de variable solo puede contener caracteres alfanuméricos y guiones bajos (Az, 0-9 y _)

* Los nombres de las variables distinguen entre mayúsculas y minúsculas (edad, Edad y EDAD son tres variables diferentes)

* Un nombre de variable no puede ser ninguna de las palabras clave de Python.

Ejemplos de nombres de variables aceptables: 

Imprimamos algunas para verificar:

In [76]:
myvar = 3
myvar

3

In [77]:
MYVAR = 4
MYVAR

4

In [78]:
myVar2 = 7
myVar2

7

In [79]:
_myvar = 9
_myvar

9

In [81]:
my___var = 8
my___var

8

Ejemplos de nombres de variables no aceptados:

In [82]:
my var = 9

SyntaxError: invalid syntax (3414335678.py, line 1)

In [84]:
my-var = 9

SyntaxError: cannot assign to operator (587289563.py, line 1)

In [85]:
4myvar = 0

SyntaxError: invalid syntax (2597150215.py, line 1)

### <span style="color:darkblue">Nombres de variables de varias palabras</span>

Los nombres de variables con más de una palabra pueden ser difíciles de leer.
Hay varias técnicas que puede utilizar para hacerlos más legibles:


### <span style="color:darkgreen">Camel Case:</span>

Se utiliza para nombrar variables en la programación donde las palabras deben estar conectadas pero aún se desea distinguir claramente las palabras individuales. 

Por ejemplo, en variables de programación en lenguajes como Java y JavaScript.

### Definición: 
    
        En Camel Case, cada palabra comienza con una letra mayúscula, excepto la primera palabra que comienza en minúscula. No se utilizan espacios ni caracteres especiales entre palabras.
        
### Ejemplo: 

```
    miVariableEjemplo

    nombreUsuario
    
    calcularImpuesto
```

### <span style="color:darkgreen">Pascal Case:</span>

 En nombres de clases y tipos de datos, como en la programación orientada a objetos. Se utiliza para nombrar clases, estructuras y tipos de datos en lenguajes como C# y Python.

### Definición: 
    
         Similar a Camel Case, pero la primera letra de cada palabra se coloca en mayúscula.
        
### Ejemplo: 

```
    MiVariableEjemplo

    NombreUsuario
    
    CrearUsuario
```

### <span style="color:darkgreen">Snake Case:</span>

Se utiliza para nombrar variables o funciones en lenguajes que son sensibles a las mayúsculas y minúsculas y que no admiten espacios en los nombres, como Python y Ruby.

### Definición: 
    
          Las palabras están separadas por guiones bajos (subrayados). Se usan solo letras minúsculas.
        
### Ejemplo: 

```
    numero_de_telefono

    cantidad_de_productos
    
    mi_variable_ejemplo
```

### <span style="color:darkgreen">Lower Case:</span>

 Se usa para nombrar variables o identificadores generales en muchos lenguajes de programación.

### Definición: 
    
          Todas las letras están en minúscula.
        
### Ejemplo: 

```
    mivariableejemplo

    cantidaddeproductos
    
    nombre
```

### <span style="color:darkgreen">Upper Case:</span>

Se usa para nombrar constantes en mayúsculas. Por ejemplo, en matemáticas, el número Pi se representa en mayúsculas para indicar que es una constante.

### Definición: 
    
          Todas las letras están en mayúscula.
        
### Ejemplo: 

```
    PI

    MAXIMO_VALOR
    
    MIVARIABLEEJEMPLO
```

### <span style="color:darkblue">Muchos valores para múltiples variables</span>

Python le permite asignar valores a múltiples variables en una línea:

In [86]:
m = 2
y = 'naranja'
z = 3.1416

print(m)
print(y)
print(z)

2
naranja
3.1416


In [87]:
m, y, z = 2, 'Naranja', 3.12416

print(m)
print(y)
print(z)

2
Naranja
3.12416


In [88]:
print(x,y,z)

3 Naranja 3.12416


In [91]:
print(x,'\n\n\t',y,'\t',z)

3 

	 Naranja 	 3.12416


### <span style="color:darkblue">Un valor para múltiples variables</span>

Y puede asignar el mismo valor a múltiples variables en una línea:

In [97]:
x= y = z = 3*'hola, mundo'
print(x, y, z)

hola, mundohola, mundohola, mundo hola, mundohola, mundohola, mundo hola, mundohola, mundohola, mundo


### <span style="color:darkblue">Salida de una variable</span>

La función `print()` de Python se usa a menudo para generar variables.


In [99]:
x = 'Python es muy sencillo y potente para la programación'
print(x)

Python es muy sencillo y potente para la programación


En la función `print()`, genera múltiples variables, separadas por una coma:

In [101]:
x = 'Python es' 
y = 'muy sencillo y potente' 
z = 'para la programación'

print(x, y, z)


Python es muy sencillo y potente para la programación


También puede usar el **operador +** para generar múltiples variables:

In [103]:
x = 'Python es' 
y = 'muy sencillo y potente ' 
z = 'para la programación'

print(x + y + z)


Python esmuy sencillo y potente para la programación


Para números, el carácter **+** funciona como un operador matemático:

In [109]:
a, b, c, = 3, 4, 5

print(a, b, c)

print(a + b + c)

print('La suma es ',a + b + c)
print('La multiplicacion es: ', a * b * c)

3 4 5
12
La suma es  12
La multiplicacion es:  60


En la función `print()`, cuando intenta combinar una cadena y un número con el operador +, Python le dará un error:

In [110]:
m, y, z = 2, 'Naranja', 3.12416

print(m + y + z)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

La mejor manera de generar múltiples variables en la función `print()` es separarlas con comas, que incluso admiten diferentes tipos de datos:

In [111]:
m, y, z = 2, 4, 3.12416

print(m + y + z)

9.12416


In [113]:
nombre   = 'Juan Francisco '
apellido = 'Lopez funes'

print('Mi nombre es: '+ nombre + 'y mi apellido es : ', apellido)

Mi nombre es: Juan Francisco y mi apellido es :  Lopez funes


### Formato de cadena f-string

In [122]:
nombre   = 'Juan'
apellido = 'Funes'
edad     = 37
horasT   = 44
pagoHora = 23.78

sueldoSemana = horasT * pagoHora

print(f'Nombre de empleado: {nombre} \n Apellido del Empleado: {apellido} \n y su salario es : {sueldoSemana : .2f }')

Nombre de empleado: Juan 
 Apellido del Empleado: Funes 
 y su salario es :  1046.32


In [132]:
print('Nombre: %s apellido: %s Salario: %.2f'%(nombre, apellido, sueldoSemana))

Nombre: Juan apellido: Funes Salario: 1046.32


In [133]:
lista_name = [' Juan', 'Mario', 'Luis', 'Iris']

In [134]:
print(lista_name)

[' Juan', 'Mario', 'Luis', 'Iris']


In [135]:
print(lista_name[0])

 Juan


In [136]:
print(lista_name[1])

Mario


In [137]:
print(lista_name[2])

Luis


In [138]:
print(lista_name[-3])

Mario


### <span style="color:darkred">Tipos de datos</span>

Python es un lenguaje de programación orientado a objetos, lo que significa que todo en Python es un objeto se convierte en tipos de datos. 

Existen varios tipos de datos incorporados que se utilizan para almacenar diferentes tipos de información. Algunos de los tipos de datos más comunes en Python son los siguientes:

#### Números

En Python, hay tres tipos de números: enteros (int), números de punto flotante (float) y números complejos (complex). Los enteros son números enteros sin decimales, como 42. 

Los números de punto flotante son números con decimales, como 3.14. 

Los números complejos son números con una parte real y una parte imaginaria, como 2 + 3j.

#### Cadenas de texto

Las cadenas de texto (str) se utilizan para almacenar texto y se definen entre comillas simples o dobles. Por ejemplo, “Hola, ¡mundo!" es una cadena de texto en Python.

#### Listas

Las listas (list) se utilizan para almacenar una colección ordenada de elementos y se definen entre corchetes. Por ejemplo, [1, 2, 3, 4] es una lista en Python.

#### Tuplas

Las tuplas (tuple) son similares a las listas, pero son inmutables, lo que significa que no se pueden modificar después de su creación. 

Las tuplas se definen entre paréntesis. Por ejemplo, (1, 2, 3, 4) es una tupla en Python.

#### Diccionarios

Los diccionarios (dict) se utilizan para almacenar una colección de pares clave-valor y se definen entre llaves. Por ejemplo, {"nombre": "Juan", "edad": 30} es un diccionario en Python.

#### Conjuntos

Los conjuntos (set) se utilizan para almacenar una colección de elementos únicos y se definen entre llaves. Por ejemplo, {1, 2, 3, 4} es un conjunto en Python.


### Tipos de datos al declarar

AL declarar una variable no es necesario indicar ningún tipo en particular, e incluso pueden cambiar de tipo después de que se hayan establecido.

In [140]:
dato_entero = 4
dato_entero

4

In [141]:
dato_flotante = 5.783478
dato_flotante

5.783478

Un número complejo o imaginario es de la forma $$a + bj$$

In [143]:
complejo = 5 + 4j
type(complejo)

complex

In [147]:
complejo_num = 3 + 0j
type(complejo_num)

complex

In [150]:
print(type(1 < 5))
print(1 < 5)

<class 'bool'>
True


In [151]:
print(type(10 < 5))
print(10 < 5)

<class 'bool'>
False


Los tipos de variable también pueden indicarse por medio de funciones que y así especificar literalmente el tipo de variable que queremos almacenar esto podrá generar una conversión de ser necesario.

Si el valor no coincide con una conversión correcta generará un error por ejemplo cambiar letras a enteras que no pueden ser cambiados. 

## Aplicando Conocimientos

### Ejercicio de aplicación:

-	Crea un archivo de Python en donde almacenaremos información personal, esta información debe ir con el tipo de dato correspondiente. 
Variables a crear:
    
o	Nombres

o	PrimerApellido

o	SegundoApellido

o	Edad

o	Nacionalidad

o	Sexo
